### PR TITLE
update to .net 4.7

### PR DESCRIPTION
**Pull request by mjkkirschner**
https://jira.autodesk.com/browse/QNTM-5552 (thought I would look at this now as it's closely related to https://github.com/DynamoDS/Dynamo/pull/9174)

updates greg's target framework to .net 4.7,
also reinstalls restSharp to use the .net 46 version of this dependency.

This PR also attempts to only hard code TLS 1.2 if we are on windows 7 - where the OS defaults for secure https protocol cannot be used.

I've tested this on windows 10, the if statement is not hit, and TLS 1.2 is still used when talking with the prod package manager.

I was under the impression that we might have to also move to a newer rest sharp that targeted .net 4.7 to get this benefit, but it seems that it's enough if the referencing assembly is built on .net47.

screen shot 2018-11-02 at 6 03 28 pm

PTAL
@alfarok @Dewb

### CODE REVIEW

**Comment by alfarok** [comment]
:+1: LGTM

File: src/GregClient/GregClient.cs [hunk 1]
```diff
@@ -23,10 +23,16 @@ public GregClient(IAuthProvider provider, string packageManagerUrl)
             // TODO can be removed when we upgrade to .net4.7
             // documented here: https://medium.com/@kyle.gagnet/your-net-code-could-stop-working-in-june-afb35fbf29ca
 
-            ServicePointManager.SecurityProtocol |= SecurityProtocolType.Tls12;
+            //https://stackoverflow.com/questions/2819934/detect-windows-version-in-net
```
Comment (QilongTang) [inline src/GregClient/GregClient.cs]: Space missing, other than this LGTM

**Comment by alfarok** [comment]
@mjkkirschner Is this good to go?

**Comment by mjkkirschner** [comment]
I think theres another task for this - I want to separate testing of this from testing of the zip/ unzip - so for now I was planning to hold off merging this until we pick up that task from the backlog.

**Comment by mjkkirschner** [comment]
@QilongTang @alfarok PTAL - I changed the nuget generation script to add the dll the net47 folder instead of net 45 to match dynamo.

File: src/GregClient/GregClient.csproj [hunk 2]
```diff
@@ -39,8 +39,7 @@
       <Private>False</Private>
     </Reference>
     <Reference Include="RestSharp, Version=105.2.3.0, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>..\packages\RestSharp.105.2.3\lib\net4\RestSharp.dll</HintPath>
-      <Private>True</Private>
+      <HintPath>..\packages\RestSharp.105.2.3\lib\net46\RestSharp.dll</HintPath>
```
Comment (QilongTang) [inline src/GregClient/GregClient.csproj]: So they have done sth similar than us and moved to a newer location?

File: src/GregClientSandbox/packages.config [hunk 3]
```diff
@@ -1,4 +1,4 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="RestSharp" version="105.2.3" targetFramework="net40" />
+  <package id="RestSharp" version="105.2.3" targetFramework="net47" />
```
Comment (QilongTang) [inline src/GregClientSandbox/packages.config]: hmmm folder is 46 but targetFramework is 47?

**Comment by mjkkirschner** [comment]
Yep

This version of restsharp only has a 46 target - remember the 47 comes from our project - the referencing assembly and nuget tries to get as close as possible in version

File: src/GregClient/GregClient.cs [hunk 4]
```diff
@@ -19,14 +19,18 @@ public IAuthProvider AuthProvider
...
-            // TODO can be removed when we upgrade to .net4.7
-            // documented here: https://medium.com/@kyle.gagnet/your-net-code-could-stop-working-in-june-afb35fbf29ca
+           
 
-            ServicePointManager.SecurityProtocol |= SecurityProtocolType.Tls12;
+            // https://stackoverflow.com/questions/2819934/detect-windows-version-in-net
+            // if the current OS is windows 7 or lower
+            // set TLS to 1.2.
+            // else do nothing and let the OS decide the version of TLS to support. (.net 4.7 required)
+            if (System.Environment.OSVersion.Version.Major <= 6 && System.Environment.OSVersion.Version.Minor <= 1)
```
Comment (alfarok) [inline src/GregClient/GregClient.cs]: the logic in that versioning chart is ridiculous! haha

**Pull request merged**